src/dune_rules/action_unexpanded.ml

```diff
@@ -415,6 +475,18 @@ let rec expand (t : Dune_lang.Action.t) ~context : Action.t Action_expander.t =
     let+ t = expand t in
     O.Ignore (outputs, t)
   | Progn l ->
+    let rec check_ppx_is_last as_ =
```

Comment:
Could you move this function the to the toplevel? We handle a lot of `Progn` constructors, so this allocation should be avoided. It also makes the code clearer.